In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from pathlib import Path

# Q1 How to read the excel starts from the 5th row?

In [2]:
data1 = pd.read_excel(
    Path.home()
    / "OneDrive"
    / "2024"
    / "Big Data Analysis"
    / "Chinese data"
    / "2024_01_17_15_34_24_N2021110004000010.xlsx",
    na_values="..",
    sheet_name="1",
    header=5   
)
data1.head()

Unnamed: 0      Unnamed: 1  Unnamed: 2  第一产业\nPrimary\nIndustry  \
0        全 国  National Total    29389255                  1827421   
1        北 京         Beijing     1174904                     6858   
2        天 津         Tianjin      371124                    10684   
3        河 北           Hebei     1456954                   106861   
4        山 西          Shanxi      772765                    97184   

   第二产业\nSecondary\nIndustry  第三产业\nTertiary\nIndustry  企业法人\nEnterprises  \
0                    5904042                  21657792           25055456   
1                      67254                   1100792            1127720   
2                      69170                    291270             341215   
3                     379431                    970662            1248654   
4                     112411                    563170             604760   

   事业法人\nInstitution\nUnits  机关法人\nGovernment\nUnits  \
0                    812962                   237879   
1                     11226                     1868   
2                      7125                     1924   
3                     32874                    11349   
4                     26563                     8339   

   社会团体\nSocial\nOrganization  其他\nOthers  
0                      334242     2948716  
1                        6047       28043  
2                        2363       18497  
3                        9091      154986  
4                        9725      123378

# Q2 How to rename columns names?

In [3]:
data1 = data1.rename(columns={"Unnamed: 0": "地区", "Unnamed: 1": "Region", "Unnamed: 2": "法人总数"}).rename(columns=lambda x: x.split("\n")[0])
data1

地区          Region      法人总数     第一产业     第二产业      第三产业      企业法人  \
0   全 国  National Total  29389255  1827421  5904042  21657792  25055456   
1   北 京         Beijing   1174904     6858    67254   1100792   1127720   
2   天 津         Tianjin    371124    10684    69170    291270    341215   
3   河 北           Hebei   1456954   106861   379431    970662   1248654   
4   山 西          Shanxi    772765    97184   112411    563170    604760   
5   内蒙古  Inner Mongolia    439523    60334    69478    309711    320329   
6   辽 宁        Liaoning    748782    50296   152619    545867    633098   
7   吉 林           Jilin    237731    26302    38739    172690    166693   
8   黑龙江    Heilongjiang    344360    53748    50982    239630    234504   
9   上 海        Shanghai    532762     5200    76033    451529    497979   
10  江 苏         Jiangsu   2540015    41935   743879   1754201   2333775   
11  浙 江        Zhejiang   2277240    48515   593845   1634880   2087888   
12  安 徽           Anhui   1160828    96705   254409    809714    998274   
13  福 建          Fujian   1156978    54003   221019    881956   1031865   
14  江 西         Jiangxi    765836    77906   155567    532363    621069   
15  山 东        Shandong   2842426   127605   666397   2048424   2483474   
16  河 南           Henan   1652264   136573   281552   1234139   1324372   
17  湖 北           Hubei   1183265    67468   225063    890734    968729   
18  湖 南           Hunan    831104    66175   138136    626793    628510   
19  广 东       Guangdong   3526206    40563   795214   2690429   3203987   
20  广 西         Guangxi    754741    81140    99060    574541    596654   
21  海 南          Hainan    140362    11535    17239    111588    117573   
22  重 庆       Chongqing    642720    78978    86191    477551    566829   
23  四 川         Sichuan    934554    93708   145612    695234    720525   
24  贵 州         Guizhou    543603   108209    92893    342501    431382   
25  云 南          Yunnan    741669    97737   105141    538791    599648   
26  西 藏           Tibet     49889     2533    12397     34959     28775   
27  陕 西         Shaanxi    689053    55017   134092    499944    540901   
28  甘 肃           Gansu    302596    61979    35020    205597    175537   
29  青 海         Qinghai    112136    18304    15464     78368     78106   
30  宁 夏         Ningxia    133774    18974    19866     94934    103540   
31  新 疆        Xinjiang    329091    24392    49869    254830    239091   

      事业法人    机关法人    社会团体       其他  
0   812962  237879  334242  2948716  
1    11226    1868    6047    28043  
2     7125    1924    2363    18497  
3    32874   11349    9091   154986  
4    26563    8339    9725   123378  
5    20501    8009    9628    81056  
6    21636    8196    6404    79448  
7    17499    5025    2702    45812  
8    20340    8948    4720    75848  
9     7769    1364    3142    22508  
10   39948    9139   32191   124962  
11   28492    7824   22336   130700  
12   22520    8507   13666   117861  
13   26700    7706   18154    72553  
14   29282    8492    9362    97631  
15   47480   12417   18876   280179  
16   70427   13874   11363   232228  
17   42109   10381   15574   146472  
18   44321   10735   12190   135348  
19   46408   13041   28965   233805  
20   44140    8900   11564    93483  
21    4164    1618    1822    15185  
22   16483    3526    7132    48750  
23   53431   15284   15591   129723  
24   25866    7322    6031    73002  
25   29454   10828   14279    87460  
26    2674    5217     436    12787  
27   28177    8364   14185    97426  
28   17733    6726   15555    87045  
29    4817    2939    4465    21809  
30    3440    1731    2281    22782  
31   19363    8286    4402    57949

# Q3 How to remove the whitespace by using str.replace? (in the first column "地区")

In [4]:
data1['地区'] = data1['地区'].str.replace(" ", "")
data1['地区'].unique()

array(['全国', '北京', '天津', '河北', '山西', '内蒙古', '辽宁', '吉林', '黑龙江', '上海', '江苏',
       '浙江', '安徽', '福建', '江西', '山东', '河南', '湖北', '湖南', '广东', '广西', '海南',
       '重庆', '四川', '贵州', '云南', '西藏', '陕西', '甘肃', '青海', '宁夏', '新疆'],
      dtype=object)

# Q4 How to add a codebook for the Chinese provinces?

In [5]:
code = pd.read_excel(
    Path.home()
    / "OneDrive"
    / "2024"
    / "Big Data Analysis"
    / "Chinese data"
    / "Codebook.xlsx",
    na_values="..",
    sheet_name="Sheet1",
    header=2,
    usecols= "B:C"    
).dropna().astype({"行政区划代码": int}).astype({"行政区划代码": str}).rename(columns={"单位名称": "地区"})

code.head()

行政区划代码      地区
0  110000     北京市
1  110101     东城区
2  110102     西城区
3  110105     朝阳区
4  110106     丰台区

## Q4.1 How to filter the provincal code?
### All provinces,cities and districts has a unique code. For instance Beijing is 110000, and the districts of Beijing are 110101, 110102, etc. So all the provincal code ends with 0000.

In [6]:
code = code[code['行政区划代码'].str[-4:] == "0000"]
code.head()

行政区划代码      地区
0    110000     北京市
17   120000     天津市
34   130000     河北省
213  140000     山西省
342  150000  内蒙古自治区

## Q4.2 How to remove part of the characters in "地区".
### The codebook "地区" contains strings such as “省”(province), "市"(city) which are different from data1. We have to move these kind of characters.

In [7]:
code["地区"] = code["地区"].str.rstrip("市省自治区壮族回族维吾尔特别行政区")
code.head()

行政区划代码   地区
0    110000   北京
17   120000   天津
34   130000   河北
213  140000   山西
342  150000  内蒙古

# Q5 How to merge two dataframes?
### By using the function of merge, be sure to understand how = "inner" or "left" or "right", and on=[""]

In [8]:
data1 = data1.merge(code, how="inner", on="地区")
data1

地区          Region     法人总数    第一产业    第二产业     第三产业     企业法人   事业法人  \
0    北京         Beijing  1174904    6858   67254  1100792  1127720  11226   
1    天津         Tianjin   371124   10684   69170   291270   341215   7125   
2    河北           Hebei  1456954  106861  379431   970662  1248654  32874   
3    山西          Shanxi   772765   97184  112411   563170   604760  26563   
4   内蒙古  Inner Mongolia   439523   60334   69478   309711   320329  20501   
5    辽宁        Liaoning   748782   50296  152619   545867   633098  21636   
6    吉林           Jilin   237731   26302   38739   172690   166693  17499   
7   黑龙江    Heilongjiang   344360   53748   50982   239630   234504  20340   
8    上海        Shanghai   532762    5200   76033   451529   497979   7769   
9    江苏         Jiangsu  2540015   41935  743879  1754201  2333775  39948   
10   浙江        Zhejiang  2277240   48515  593845  1634880  2087888  28492   
11   安徽           Anhui  1160828   96705  254409   809714   998274  22520   
12   福建          Fujian  1156978   54003  221019   881956  1031865  26700   
13   江西         Jiangxi   765836   77906  155567   532363   621069  29282   
14   山东        Shandong  2842426  127605  666397  2048424  2483474  47480   
15   河南           Henan  1652264  136573  281552  1234139  1324372  70427   
16   湖北           Hubei  1183265   67468  225063   890734   968729  42109   
17   湖南           Hunan   831104   66175  138136   626793   628510  44321   
18   广东       Guangdong  3526206   40563  795214  2690429  3203987  46408   
19   广西         Guangxi   754741   81140   99060   574541   596654  44140   
20   海南          Hainan   140362   11535   17239   111588   117573   4164   
21   重庆       Chongqing   642720   78978   86191   477551   566829  16483   
22   四川         Sichuan   934554   93708  145612   695234   720525  53431   
23   贵州         Guizhou   543603  108209   92893   342501   431382  25866   
24   云南          Yunnan   741669   97737  105141   538791   599648  29454   
25   西藏           Tibet    49889    2533   12397    34959    28775   2674   
26   陕西         Shaanxi   689053   55017  134092   499944   540901  28177   
27   甘肃           Gansu   302596   61979   35020   205597   175537  17733   
28   青海         Qinghai   112136   18304   15464    78368    78106   4817   
29   宁夏         Ningxia   133774   18974   19866    94934   103540   3440   
30   新疆        Xinjiang   329091   24392   49869   254830   239091  19363   

     机关法人   社会团体      其他  行政区划代码  
0    1868   6047   28043  110000  
1    1924   2363   18497  120000  
2   11349   9091  154986  130000  
3    8339   9725  123378  140000  
4    8009   9628   81056  150000  
5    8196   6404   79448  210000  
6    5025   2702   45812  220000  
7    8948   4720   75848  230000  
8    1364   3142   22508  310000  
9    9139  32191  124962  320000  
10   7824  22336  130700  330000  
11   8507  13666  117861  340000  
12   7706  18154   72553  350000  
13   8492   9362   97631  360000  
14  12417  18876  280179  370000  
15  13874  11363  232228  410000  
16  10381  15574  146472  420000  
17  10735  12190  135348  430000  
18  13041  28965  233805  440000  
19   8900  11564   93483  450000  
20   1618   1822   15185  460000  
21   3526   7132   48750  500000  
22  15284  15591  129723  510000  
23   7322   6031   73002  520000  
24  10828  14279   87460  530000  
25   5217    436   12787  540000  
26   8364  14185   97426  610000  
27   6726  15555   87045  620000  
28   2939   4465   21809  630000  
29   1731   2281   22782  640000  
30   8286   4402   57949  650000

# Q6 How to generate a new columns with a constant value?

In [9]:
data1['年份'] = 2020

In [10]:
data2= pd.read_excel(
    Path.home() / "OneDrive" / "2024" / "Big Data Analysis" / "Chinese data"/ "论文数据.xlsx",
    sheet_name="全部数据",
).merge(code, how="inner", on="地区")
data2.head()

序号  地区    年份  国内生产总值（当年价）（亿元）  第三产业增加值（亿元）  总人口（万人）  城镇人口（万人）  城镇就业人员数（万人）  \
0   1  北京  2013         19800.81     14986.43     2125      1836    1162.7905   
1   1  北京  2014         21330.83     16627.04     2171      1878    1309.2000   
2   1  北京  2015         23014.59     18331.74     2188      1897    1415.7448   
3   1  北京  2016         25669.13     20594.90     2195      1904    1476.6700   
4   1  北京  2017         28014.94     22567.76     2194      1907    1547.1100   

   货物出口金额（万美元）  货物进口金额（万美元）  一般预算支出（亿元）  互联网宽带接入用户（万户）  有电子商务交易活动企业占总企业数比重（%）  \
0  6309756.100  36589825.10   4173.6563        480.400                 7.4626   
1  6233842.000  35318018.00   4524.6700        482.400                12.5790   
2  5466681.700  26477375.20   5737.7000        491.900                17.1000   
3  5202284.300  23032611.70   6406.7700        475.760                18.0000   
4  5850304.551  26521615.92   6824.5284        541.936                19.0000   

   电子商务销售额（亿元）  电子商务采购额（亿元）  R&D经费（万元）   数字普惠金融指数  行政区划代码  
0    7467.6372    4877.7164  2130617.5  229.57001  110000  
1    9012.4132    7563.9099  2335010.1  235.22000  110000  
2   10530.5028    9534.1257  2440874.5  379.48001  110000  
3   12026.6910    9340.6360  2548433.3  329.89999  110000  
4   18385.6627   11055.8269  2690851.2  326.01999  110000

In [11]:
df = data1.drop(columns=['地区']).merge(data2, how="outer", on=["行政区划代码","年份"]).sort_values(["行政区划代码","年份"])
df.to_csv(
    Path.home() / "OneDrive" / "2024" / "Big Data Analysis" /"Chinese data"/ "df.csv",
    index=True,
)

# Let's do some simple calculations!

In [12]:
### Check whether there are missing values.
data2.isna().sum().sort_values(ascending=False)

序号                       0
地区                       0
数字普惠金融指数                 0
R&D经费（万元）                0
电子商务采购额（亿元）              0
电子商务销售额（亿元）              0
有电子商务交易活动企业占总企业数比重（%）    0
互联网宽带接入用户（万户）            0
一般预算支出（亿元）               0
货物进口金额（万美元）              0
货物出口金额（万美元）              0
城镇就业人员数（万人）              0
城镇人口（万人）                 0
总人口（万人）                  0
第三产业增加值（亿元）              0
国内生产总值（当年价）（亿元）          0
年份                       0
行政区划代码                   0
dtype: int64

# Q7 How to interpolate the missing values?

In [13]:
### Interpolate
### https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.interpolate.html

df = data2.interpolate(method='linear', limit_direction='both')
#df.isna().sum().sort_values(ascending=False)

# Q8 How to add a new column based on the simple calculations such as sum, divide, multiply, ln?

In [14]:
df.columns

Index(['序号', '地区', '年份', '国内生产总值（当年价）（亿元）', '第三产业增加值（亿元）', '总人口（万人）',
       '城镇人口（万人）', '城镇就业人员数（万人）', '货物出口金额（万美元）', '货物进口金额（万美元）', '一般预算支出（亿元）',
       '互联网宽带接入用户（万户）', '有电子商务交易活动企业占总企业数比重（%）', '电子商务销售额（亿元）', '电子商务采购额（亿元）',
       'R&D经费（万元）', '数字普惠金融指数', '行政区划代码'],
      dtype='object')

In [15]:
df['import_export'] = df['货物出口金额（万美元）'] + df['货物进口金额（万美元）']

df['internet'] = df['互联网宽带接入用户（万户）'] / df['总人口（万人）']

### https://numpy.org/doc/stable/reference/generated/numpy.log.html
df['lngdp'] = np.log(df["国内生产总值（当年价）（亿元）"])

In [21]:
df.head()

序号  地区    年份  国内生产总值（当年价）（亿元）  第三产业增加值（亿元）  总人口（万人）  城镇人口（万人）  城镇就业人员数（万人）  \
0   1  北京  2013         19800.81     14986.43     2125      1836    1162.7905   
1   1  北京  2014         21330.83     16627.04     2171      1878    1309.2000   
2   1  北京  2015         23014.59     18331.74     2188      1897    1415.7448   
3   1  北京  2016         25669.13     20594.90     2195      1904    1476.6700   
4   1  北京  2017         28014.94     22567.76     2194      1907    1547.1100   

   货物出口金额（万美元）  货物进口金额（万美元）  ...  电子商务销售额（亿元）  电子商务采购额（亿元）  R&D经费（万元）  \
0  6309756.100  36589825.10  ...    7467.6372    4877.7164  2130617.5   
1  6233842.000  35318018.00  ...    9012.4132    7563.9099  2335010.1   
2  5466681.700  26477375.20  ...   10530.5028    9534.1257  2440874.5   
3  5202284.300  23032611.70  ...   12026.6910    9340.6360  2548433.3   
4  5850304.551  26521615.92  ...   18385.6627   11055.8269  2690851.2   

    数字普惠金融指数  行政区划代码  import_export  internet      lngdp  R&Dgrowthrate  \
0  229.57001  110000   4.289958e+07  0.226071   9.893478            NaN   
1  235.22000  110000   4.155186e+07  0.222202   9.967909       0.095931   
2  379.48001  110000   3.194406e+07  0.224817  10.043884       0.045338   
3  329.89999  110000   2.823490e+07  0.216747  10.153044       0.044066   
4  326.01999  110000   3.237192e+07  0.247008  10.240493       0.055884   

         sell  
0   7467.6372  
1  16480.0504  
2  27010.5532  
3  39037.2442  
4  57422.9069  

[5 rows x 23 columns]

# Q9 How to calculate the growth rate of a certain column?
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.pct_change.html

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html


In [17]:
df['R&Dgrowthrate'] = df.groupby('行政区划代码')['R&D经费（万元）'].pct_change()

In [18]:
df.query('年份==2013')

序号   地区    年份  国内生产总值（当年价）（亿元）  第三产业增加值（亿元）  总人口（万人）  城镇人口（万人）  \
0     1   北京  2013         19800.81     14986.43     2125      1836   
10    2   天津  2013         14442.01      6905.03     1410      1160   
20    3   河北  2013         28442.95     10038.89     7288      3500   
30    4   山西  2013         12665.25      5035.75     3535      1869   
40    5  内蒙古  2013         16916.50      6148.78     2455      1469   
50    6   辽宁  2013         27213.22     10486.56     4365      2901   
60    7   吉林  2013         13046.40      4613.89     2668      1487   
70    8  黑龙江  2013         14454.91      5947.92     3666      2128   
80    9   上海  2013         21818.15     13445.07     2448      2193   
90   10   江苏  2013         59753.37     26421.64     8192      5275   
100  11   浙江  2013         37756.58     17337.22     5784      3698   
110  12   安徽  2013         19229.34      6286.82     5988      2866   
120  13   福建  2013         21868.49      8508.03     3885      2362   
130  14   江西  2013         14410.19      5030.63     4476      2195   
140  15   山东  2013         55230.32     22519.23     9746      5210   
150  16   河南  2013         32191.30     10290.49     9573      4174   
160  17   湖北  2013         24791.83      9398.77     5798      3160   
170  18   湖南  2013         24621.67      9885.09     6600      3144   
180  19   广东  2013         62474.79     29688.97    11270      7674   
190  20   广西  2013         14449.90      5171.39     4731      2134   
200  21   海南  2013          3177.56      1518.70      920       481   
210  22   重庆  2013         12783.26      5242.03     3011      1755   
220  23   四川  2013         26392.07      9256.13     8109      3646   
230  24   贵州  2013          8086.86      3734.04     3632      1376   
240  25   云南  2013         11832.31      4897.75     4641      1856   
250  26   西藏  2013           815.67       427.93      317        76   
260  27   陕西  2013         16205.45      5607.52     3804      1962   
270  28   甘肃  2013          6330.69      2567.60     2537      1027   
280  29   青海  2013          2122.06       689.15      571       281   
290  30   宁夏  2013          2577.57      1077.12      666       352   
300  31   新疆  2013          8443.84      3125.98     2285      1027   

     城镇就业人员数（万人）  货物出口金额（万美元）  货物进口金额（万美元）  ...  有电子商务交易活动企业占总企业数比重（%）  \
0    1162.790500    6309756.1   36589825.1  ...                 7.4626   
10    434.811100    4900493.8    7949685.0  ...                 5.3837   
20   1083.518600    3096060.9    2395096.2  ...                 2.6886   
30    696.133000     799557.4     779540.5  ...                 1.7619   
40    645.306270     409256.1     790201.1  ...                 1.9638   
50   1301.783000    6452200.7    4995618.2  ...                 2.0058   
60    699.535520     673890.6    1909282.9  ...                 1.7217   
70    860.100710    1623172.6    2264736.8  ...                 1.5007   
80   1045.130000   20418002.6   23708819.0  ...                 5.7054   
90   3181.800000   32880175.2   22200051.8  ...                 5.5930   
100  2080.078900   24874624.1    8704246.4  ...                11.5456   
110  1043.566300    2825131.4    1726765.5  ...                 4.7494   
120  1129.808200   10647442.0    6284648.1  ...                 5.7585   
130   813.374570    2816665.2     857998.2  ...                 3.2374   
140  2052.587200   13419012.5   13234140.8  ...                 4.5671   
150  1535.168700    3598710.2    2396976.6  ...                 2.8133   
160  1319.526200    2283620.8    1354386.8  ...                 3.3758   
170  1253.929300    1482119.6    1035411.2  ...                 4.0134   
180  3637.120600   63636384.5   45521759.2  ...                 7.7019   
190   700.668580    1869325.7    1413424.2  ...                 3.7460   
200   217.848010     370648.6    1127894.8  ...                 4.7898   
210   907.749570    4679589.9    2189626.5  ...                 3.4803   
220  1360.256500    4194905.6 

# Q10 How to calcualte the cumulative sum?
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cumsum.html

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.groupby.html


In [19]:
df['sell'] = df.groupby('行政区划代码')['电子商务销售额（亿元）'].cumsum()

In [20]:
df.head()

序号  地区    年份  国内生产总值（当年价）（亿元）  第三产业增加值（亿元）  总人口（万人）  城镇人口（万人）  城镇就业人员数（万人）  \
0   1  北京  2013         19800.81     14986.43     2125      1836    1162.7905   
1   1  北京  2014         21330.83     16627.04     2171      1878    1309.2000   
2   1  北京  2015         23014.59     18331.74     2188      1897    1415.7448   
3   1  北京  2016         25669.13     20594.90     2195      1904    1476.6700   
4   1  北京  2017         28014.94     22567.76     2194      1907    1547.1100   

   货物出口金额（万美元）  货物进口金额（万美元）  ...  电子商务销售额（亿元）  电子商务采购额（亿元）  R&D经费（万元）  \
0  6309756.100  36589825.10  ...    7467.6372    4877.7164  2130617.5   
1  6233842.000  35318018.00  ...    9012.4132    7563.9099  2335010.1   
2  5466681.700  26477375.20  ...   10530.5028    9534.1257  2440874.5   
3  5202284.300  23032611.70  ...   12026.6910    9340.6360  2548433.3   
4  5850304.551  26521615.92  ...   18385.6627   11055.8269  2690851.2   

    数字普惠金融指数  行政区划代码  import_export  internet      lngdp  R&Dgrowthrate  \
0  229.57001  110000   4.289958e+07  0.226071   9.893478            NaN   
1  235.22000  110000   4.155186e+07  0.222202   9.967909       0.095931   
2  379.48001  110000   3.194406e+07  0.224817  10.043884       0.045338   
3  329.89999  110000   2.823490e+07  0.216747  10.153044       0.044066   
4  326.01999  110000   3.237192e+07  0.247008  10.240493       0.055884   

         sell  
0   7467.6372  
1  16480.0504  
2  27010.5532  
3  39037.2442  
4  57422.9069  

[5 rows x 23 columns]